cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i

```diff
@@ -220,26 +291,320 @@ services::Status PredictRegressionTask<algorithmFPType, cpu>::run(const gbt::reg
+ * \brief Recursive function that computes the feature attributions for a single tree.
+ * \param iTree index of current tree
+ * \param feat dense feature vector, if the feature is missing the field is set to NaN
+ * \param phi dense output vector of feature attributions
+ * \param nFeatures number features, i.e. length of feat and phi vectors
+ * \param nodeIndex the index of the current node in the tree, counted from 1
+ * \param depth how deep are we in the tree
+ * \param uniqueDepth how many unique features are above the current node in the tree
+ * \param parentUniquePath a vector of statistics about our current path through the tree
+ * \param parentZeroFraction what fraction of the parent path weight is coming as 0 (integrated)
+ * \param parentOneFraction what fraction of the parent path weight is coming as 1 (fixed)
+ * \param parentFeatureIndex what feature the parent node used to split
+ * \param condition fix one feature to either off (-1) on (1) or not fixed (0 default)
+ * \param conditionFeature the index of the feature to fix
+ * \param conditionFraction what fraction of the current weight matches our conditioning feature
+ */
+template <typename algorithmFPType, CpuType cpu>
+template <bool hasUnorderedFeatures, bool hasAnyMissing>
+void PredictRegressionTask<algorithmFPType, cpu>::treeShap(size_t iTree, const algorithmFPType * x, algorithmFPType * phi, size_t nFeatures,
+                                                           size_t nodeIndex, size_t depth, std::uint32_t uniqueDepth, PathElement * parentUniquePath,
+                                                           float parentZeroFraction, float parentOneFraction,
+                                                           gbt::prediction::internal::FeatureIndexType parentFeatureIndex, int condition,
+                                                           gbt::prediction::internal::FeatureIndexType conditionFeature,
+                                                           float conditionFraction) const
+{
+    const gbt::internal::GbtDecisionTree * tree = _aTree[iTree];
+
+    const size_t nNodes = tree->getNumberOfNodes();
+    // splitValues contain
+    //   - the feature value that is used for the split for internal nodes
+    //   - the tree prediction for leaf nodes
+    // we are accounting for the fact that nodeIndex is counted from 1 (not from 0) as required by helper functions
+    const gbt::prediction::internal::ModelFPType * const splitValues     = tree->getSplitPoints() - 1;
+    const gbt::prediction::internal::FeatureIndexType * const fIndexes   = tree->getFeatureIndexesForSplit() - 1;
+    const gbt::prediction::internal::ModelFPType * const nodeCoverValues = tree->getNodeCoverValues() - 1;
+    const int * const defaultLeft                                        = tree->getDefaultLeftForSplit() - 1;
+
+    // stop if we have no weight coming down to us
+    if (conditionFraction == 0) return;
```

Comment:
1) Not a great idea to compare floating point variable with zero. Please compare with machine epsilon
2) You can check it right at the beginning of the function. It can bring (very) minor performance improvement